Sources/GRPC/ConnectionManager.swift

```diff
@@ -0,0 +1,586 @@
+
+      // We're mid-connection: the application doesn't have any 'ready' channels so we'll succeed
+      // the shutdown future and deal with any fallout from the connecting channel without the
+      // application knowing.
+      case .connecting(let state):
+        shutdown = ShutdownState(closeFuture: self.eventLoop.makeSucceededFuture(()))
+        self.state = .shutdown(shutdown)
+
+        // Fail the ready channel promise: we're shutting down so even if we manage to successfully
+        // connect the application shouldn't should have access to the channel.
+        state.readyChannelPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+        // In case we do successfully connect, close immediately.
+        state.candidate.whenSuccess {
+          $0.close(mode: .all, promise: nil)
+        }
+
+      // We have an active channel but the application doesn't know about it yet. We'll do the same
+      // as for `.connecting`.
+      case .active(let state):
+        shutdown = ShutdownState(closeFuture: self.eventLoop.makeSucceededFuture(()))
+        self.state = .shutdown(shutdown)
+
+        // Fail the ready channel promise: we're shutting down so even if we manage to successfully
+        // connect the application shouldn't should have access to the channel.
+        state.readyChannelPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+        // We have a channel, close it.
+        state.candidate.close(mode: .all, promise: nil)
+
+      // The channel is up and running: the application could be using it. We can close it and
+      // return the `closeFuture`.
+      case .ready(let state):
+        shutdown = ShutdownState(closeFuture: state.channel.closeFuture)
+        self.state = .shutdown(shutdown)
+
+        // We have a channel, close it.
+        state.channel.close(mode: .all, promise: nil)
+
+      // Like `.connecting` and `.active` the application does not have a `.ready` channel. We'll
+      // do the same but also cancel any scheduled connection attempts and deal with any fallout
+      // if we cancelled too late.
+      case .transientFailure(let state):
+        // Stop the creation of a new channel, if we can.
+        state.scheduled.cancel()
+        shutdown = ShutdownState(closeFuture: self.eventLoop.makeSucceededFuture(()))
+        self.state = .shutdown(shutdown)
+
+        // Fail the ready channel promise: we're shutting down so even if we manage to successfully
+        // connect the application shouldn't should have access to the channel.
+        state.readyChannelPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+
+      // We're already shutdown; nothing to do.
+      case .shutdown(let state):
+        shutdown = state
+      }
+
+      return shutdown.closeFuture
+    }
+  }
+
+  // MARK: - State changes from the channel handler.
+
+  /// The connecting channel became `active`. Must be called on the `EventLoop`.
+  internal func channelActive(channel: Channel) {
```

Comment:
Aren't we supposed to close the channel here?

Reply:
Why?

Reply:
If we're in `shutdown` and we get `channelActive`, who is closing it?

Reply:
Ah, yes, we should!